<img width="1075" alt="image" src="https://github.com/user-attachments/assets/4ccba14a-5e75-41ba-bfd9-c77ef5140fda" />


```
northwood@system76-pc:~/development/shaur/linux-socfpga$ nix develop .# -c fish
================================================
==         Caveat Lector                      ==
================================================
.config for the  located at /nix/store/lrm7xvyj9xn7lm9kk05mgrvc5yxw4fgd-linux-config-aarch64-unknown-linux-gnu-6.6.22
If you would like to use this, then run just link-config
Welcome to fish, the friendly interactive shell
Type help for instructions on how to use fish
northwood@system76-pc ~/d/s/linux-socfpga (devshell)> just link-config
ln $NIX_CONFIGFILE .config
ln: failed to create hard link '.config' => '/nix/store/lrm7xvyj9xn7lm9kk05mgrvc5yxw4fgd-linux-config-aarch64-unknown-linux-gnu-6.6.22': Operation not permitted
error: Recipe `link-config` failed on line 20 with exit code 1
northwood@system76-pc ~/d/s/linux-socfpga (devshell) [1]> just menuconfig
make menuconfig
  HOSTCC  scripts/kconfig/mconf.o
In file included from scripts/kconfig/mconf.c:23:
scripts/kconfig/lxdialog/dialog.h:19:10: fatal error: ncurses.h: No such file or directory
   19 | #include <ncurses.h>
      |          ^~~~~~~~~~~
compilation terminated.
make[2]: *** [scripts/Makefile.host:131: scripts/kconfig/mconf.o] Error 1
make[1]: *** [/home/northwood/development/shaur/linux-socfpga/Makefile:697: menuconfig] Error 2
make: *** [Makefile:234: __sub-make] Error 2
error: Recipe `menuconfig` failed on line 25 with exit code 2
northwood@system76-pc ~/d/s/linux-socfpga (devshell) [2]> just build
make -j$(nproc) $makeFlags
***
*** Configuration file ".config" not found!
***
*** Please run some configurator (e.g. "make oldconfig" or
*** "make menuconfig" or "make xconfig").
***
/home/northwood/development/shaur/linux-socfpga/Makefile:775: include/config/auto.conf.cmd: No such file or directory
make[1]: *** [/home/northwood/development/shaur/linux-socfpga/Makefile:784: .config] Error 1
make: *** [Makefile:234: __sub-make] Error 2
error: Recipe `build` failed on line 8 with exit code 2
northwood@system76-pc ~/d/s/linux-socfpga (devshell) [2]>
```

unable to run commands in fish shell or bash shell (link,build,menu, etc...) did this work for you in the dump above?

--- a/NORTHWOOD.md
+++ b/NORTHWOOD.md
@@ -1,0 +1,31 @@
+# Northwood Altera Linux
+
+## Developer Environment
+```bash
+$ nix develop .#
+```
+
+This will drop you into a shell with all of the required tools to compile the Altera fork of the linux kernel.
+
+
+### menuconfig
+Uncomment out `# menuconfig = true` in the `flake.nix` and reenter your shell with `nix develop .#`
+
+### xconfig
+Uncomment out `# xconfig = true` in the `flake.nix` and reenter your shell with `nix develop .#`
+
+## Building
+Enter the shell and see supported aliases using just.
+
+```bash
+northwood@build4:~/linux-socfpga$ just
+Available recipes:
+    default     # Display this message
+
+    [Build]
+    build       # Build Linux Kernel Image for Northwood Beamformer
+    clean       # Removes build artifacts and generated configuration files
+    link-config # Link the nix-build generated linux config to .config
+    menuconfig  # Configure linux kernel parameters via menuconfig. You will need to be in the shell with menuconfig = true; in the flake.nix
+    xconfig     # Configure linux kernel parameters via xconfig. You will need to be in the shell with xconfig = true; in the flake.nix
+```

--- a/NORTHWOOD.md
+++ b/NORTHWOOD.md
@@ -1,5 +1,7 @@
 # Northwood Altera Linux
 
+See [Building Linux Kernel](https://www.rocketboards.org/foswiki/Documentation/BuildingBootloaderStratix10#Building_Linux_Kernel) from the Building Bootloader Stratix10 tutorial provided by Altera.
+
 ## Developer Environment
 ```bash
 $ nix develop .#

--- a/flake.nix
+++ b/flake.nix
@@ -20,6 +20,7 @@
 
             # menuconfig = true;
             # xconfig = true;
+            fit-generation = true;
           };
         }
       );

--- a/flake.nix
+++ b/flake.nix
@@ -2,7 +2,7 @@
   description = "A very basic flake for building altera linux";
 
   inputs = {
-    northwood-nixpkgs.url = "git+ssh://git@github.com/Northwood-Space/northwood-nixpkgs?ref=main&rev=4388c8560e03ce7b45b74c49245a61d719a64fbe";
+    northwood-nixpkgs.url = "git+ssh://git@github.com/Northwood-Space/northwood-nixpkgs?ref=main&rev=78340900e0862efaf0b2d23aa55a1155327cd78c";
     flake-utils.url = "github:numtide/flake-utils";
   };
 

--- a/justfile
+++ b/justfile
@@ -18,7 +18,12 @@ clean:
 [group('Build')]
 link-config:
   ln -sf $NIX_CONFIGFILE .config
-  
+
+# Link the nix-build generated initrd to our local path
+[group('Build')]
+link-initrd:
+  ln -sf $INITRD initrd.zst
+
 # Configure linux kernel parameters via menuconfig. You will need to be in the shell with menuconfig = true; in the flake.nix
 [group('Build')]
 menuconfig:

--- a/justfile
+++ b/justfile
@@ -17,7 +17,7 @@ clean:
 # Link the nix-build generated linux config to .config
 [group('Build')]
 link-config:
-  ln $NIX_CONFIGFILE .config
+  ln -sf $NIX_CONFIGFILE .config
   
 # Configure linux kernel parameters via menuconfig. You will need to be in the shell with menuconfig = true; in the flake.nix
 [group('Build')]

--- a/justfile
+++ b/justfile
@@ -28,3 +28,12 @@ menuconfig:
 [group('Build')]
 xconfig:
   make xconfig
+
+# Build a Flatted uImage Tree (FIT) kernel. Requires Image, initrd.zst, and beamformer.dtb to be at the root.
+[group('Build')]
+fit-image:
+  #!/usr/bin/env bash
+
+  set -euox pipefail
+
+  mkimage -f $FIT_CONFIG kernel.itb

--- a/shell.nix
+++ b/shell.nix
@@ -55,7 +55,7 @@ let
 
     '';
     NIX_CONFIGFILE=configfile;
-    INITRD=initrd;
+    INITRD="${initrd}/initrd.zst";
     FIT_CONFIG=fit-config;
   });
 in

--- a/shell.nix
+++ b/shell.nix
@@ -30,6 +30,7 @@ let
   # https://ryantm.github.io/nixpkgs/using/overrides/#sec-pkg-overrideAttrs
   configfile = pkgsCross.altera-linux.configfile;
   initrd = if fit-generation then pkgsCross.beamformer.netbootRamdisk else "";
+  fit-config = if fit-generation then pkgsBuild.beamformer-fit-config else "";
   drv = pkgsCross.altera-linux.overrideAttrs(final: prev: {
     # give it a new name so we can differentiate between nix build derivations and nix shell roots
     pname = "northwood-altera-linux";
@@ -45,14 +46,17 @@ let
       echo "================================================"
       echo "==         Caveat Lector                      =="
       echo "================================================"
-      echo "This shell pulls in a few extraneous dependencies\n"
+      echo "This shell pulls in a few extraneous dependencies"
       echo "The linux configfile is stored in $NIX_CONFIGFILE and can be copied locally using `just link-config`"
     '' + lib.strings.optionalString fit-generation ''
       echo "Initrd stored in \$INITRD environment variable."
+      echo "Fit Image stored in \$FIT_CONFIG"
       echo "See the buildPhase of northwood-nixpkgs/pkgs/machines/beamformer/generate-fit.nix for building image"
+
     '';
     NIX_CONFIGFILE=configfile;
     INITRD=initrd;
+    FIT_CONFIG=fit-config;
   });
 in
 drv

--- a/shell.nix
+++ b/shell.nix
@@ -4,8 +4,10 @@
 # Optionally add dependencies for xconfig and menu config 
 , menuconfig ? false
 , xconfig ? false
+, fit-generation ? false
 }:
 let
+  lib = pkgsBuild.lib;
   # If we are aarch64-linux, then we do not need a cross-toolchain.
   # Otherwise, grab the pkgsCross
   #
@@ -21,11 +23,13 @@ let
   pkgsBuild = pkgsCross.buildPackages;
   menuconfigAttrs = if menuconfig then [ pkgsBuild.pkg-config pkgsBuild.ncurses ] else [];
   xconfigAttrs = if xconfig then [ pkgsBuild.pkg-config pkgsBuild.qt5.qtbase ] else [];
+  fitGenerationAttrs = if fit-generation then [ pkgsBuild.zstd pkgsBuild.xz pkgsBuild.ubootTools ] else [];
   # We want to use the same build environment that our nix derivation uses but with some new friends.
   #
   # This is why we choose to use overrideAttrs
   # https://ryantm.github.io/nixpkgs/using/overrides/#sec-pkg-overrideAttrs
   configfile = pkgsCross.altera-linux.configfile;
+  initrd = if fit-generation then pkgsCross.beamformer.netbootRamdisk else "";
   drv = pkgsCross.altera-linux.overrideAttrs(final: prev: {
     # give it a new name so we can differentiate between nix build derivations and nix shell roots
     pname = "northwood-altera-linux";
@@ -35,15 +39,20 @@ let
       # add our optional inputs
       menuconfigAttrs
       xconfigAttrs
+      fitGenerationAttrs
     ];
     shellHook = ''
       echo "================================================"
       echo "==         Caveat Lector                      =="
       echo "================================================"
-      echo ".config for the $NIX_BUILD located at ${configfile}"
-      echo "If you would like to use this, then run just link-config"
+      echo "This shell pulls in a few extraneous dependencies\n"
+      echo "The linux configfile is stored in $NIX_CONFIGFILE and can be copied locally using `just link-config`"
+    '' + lib.strings.optionalString fit-generation ''
+      echo "Initrd stored in \$INITRD environment variable."
+      echo "See the buildPhase of northwood-nixpkgs/pkgs/machines/beamformer/generate-fit.nix for building image"
     '';
     NIX_CONFIGFILE=configfile;
+    INITRD=initrd;
   });
 in
 drv

